File: 4th/homework_3/id_69/122-BestTimeToBuyAndSellStockII.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/description/
# Say you have an array for which the ith element is the price of a given stock on day i.

# Design an algorithm to find the maximum profit.
# You may complete as many transactions as you like
# (ie, buy one and sell one share of the stock multiple times).
# However, you may not engage in multiple transactions at the same time
# (ie, you must sell the stock before you buy again).
# 1, 4,3 或则 4，3，2 或则1，2，3
# 每次涨价都卖出
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        return sum(max(prices[i + 1] - prices[i], 0) for i in range(len(prices) - 1))

File: 4th/homework_3/id_45/122_best_time_to_buy_and_sell_stockII.cpp
//To make max profit
//just need to buy at the lowest point of monotone increasing
//and sell at the highest point of monotone increasing.
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int result = 0;
        for (int i = 0; i < n - 1; ++i) {
            if (prices[i+1] - prices[i] > 0)
                result += prices[i+1] - prices[i];
        }
        return result;
    }
};
